enumerable_custom.rb

```diff
@@ -0,0 +1,126 @@
+# rubocop:disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+def check_all_args(arg)
+  if arg.class == Regexp
+    my_each { |item| return false if arg.match?(item.to_s) == false }
+  elsif arg.class == Class
+    my_each { |item| return false if item.is_a?(arg) == false }
+  else
+    my_all? { |item| item == arg }
+  end
+end
+
+module Enumerable
+  def my_each
+    return enum_for unless block_given?
+
+    i = 0
+    array = to_a
+    while i < array.length
+      yield(array[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    array = to_a
+    i = 0
+    until i == array.length
+      yield(array[i], i)
+      i += 1
+    end
+  end
+
+  def my_select(*)
+    return enum_for unless block_given?
+
+    array = []
+    my_each { |item| array << item if yield(item) }
+    array
+  end
+
+  def my_all?(arg = nil)
+    return check_all_args(arg) unless arg.nil?
+
+    condition = true
+    if block_given?
+      my_each { |item| condition = false if yield(item) == false }
+    else
+      my_each { |item| condition = false if item.nil? || item == false }
+    end
+    condition
+  end
+
+  def my_any?(arg = nil)
+    condition = false
+    if block_given?
+      my_each { |item| condition = true if yield(item) == true }
+    elsif arg.is_a?(Class)
+      my_each { |item| condition = true if item.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      my_each { |item| condition = true if arg.match?(item.to_s) == true }
+    elsif arg.nil? == false
+      my_each { |item| condition = true if item == arg }
+    else
+      my_each { |item| condition = true if item }
+    end
+    condition
+  end
+
+  def my_none?(arg = nil, &proc)
+    !my_any?(arg, &proc)
+  end
+
+  def my_count(*args)
+    array = is_a?(Range) ? to_a : self
+    return array.length if !block_given? && args.empty?
+
+    number = 0
+    if block_given?
+      num_elements = []
+      array.my_each do |item|
+        if yield(item)
+          num_elements << item
+          number = num_elements.length
+        end
+      end
+    else
+      array.my_each do |item|
+        unless args.empty?
+          number += 1 if item == args[0]
+        end
+      end
+    end
+    number
+  end
+
+  def my_map(proc = nil)
+    return to_enum(:my_map) unless block_given?
+
+    map = []
+    my_each { |item| map << proc.call(item) }
+    map
+  end
```

Comment:
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns a new array with the results of running block once for every element in enum
```ruby
block = proc { |num| num < (1 + 10) / 2 }
[1, 2 ,3].my_map(&block)   #should return [1, 2, 3]
[6, 7 ,5].my_map { |num| num > 5 }   #should also return [6, 7]
```